### ISSUE TITLE
Trie alternation with empty group at start of one alternative fails

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#10721]
Migrated from [rt.perl.org#78356](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=78356) (status was 'resolved')

Searchable as RT78356$

**Comment by p5pRT** [comment]
### From @ilmari

#### Created by @ilmari



$ perl5\.8\.8 \-le 'print "foo" =~ /\\A\(?&#8203;:\(?&#8203;:\)foo|bar|zot\)\\z/ ? "match" : "FAIL"'
match

$ perl5\.12\.2 \-le 'print "foo" =~ /\\A\(?&#8203;:\(?&#8203;:\)foo|bar|zot\)\\z/ ? "match" : "FAIL"'
FAIL

Removing any of the the anchors\, the empty group\, or one of the
alterations makes it match\, as does changing any of the other
alternatives to starting with "f"\.

This regression was introduced in perl 5\.9\.5\.

$ perl5\.12\.2 \-le 'use re Debug => "ALL"; print "foo" =~ /\\A\(?&#8203;:\(?&#8203;:\)foo|bar|zot\)\\z/ ? "match" : "FAIL"'

Compiling REx "\\A\(?&#8203;:\(?&#8203;:\)foo|bar|zot\)\\z"
Starting first pass \(sizing\)
 >\\A\(?&#8203;:\(?&#8203;:\)f\.\.\. |   1|  reg    
&nbsp;                |    |    brnc   
&nbsp;                |    |      piec   
&nbsp;                |    |        atom   
 >\(?&#8203;:\(?&#8203;:\)foo\.\.\. |   2|      piec   
&nbsp;                |    |        atom   
 >?&#8203;:\(?&#8203;:\)foo|\.\.\. |    |          reg    
 >\(?&#8203;:\)foo|ba\.\.\. |    |            brnc   
&nbsp;                |    |              piec   
&nbsp;                |    |                atom   
 >?&#8203;:\)foo|bar\.\.\. |    |                  reg    
 >\)foo|bar|z\.\.\. |    |                    brnc   
 >foo|bar|zo\.\.\. |   3|              piec   
&nbsp;                |    |                atom   
 >|bar|zot\)\\\.\.\. |   5|            inst \- BRANCH
 >bar|zot\)\\z\<   |   6|            brnc   
&nbsp;                |   7|              piec   
&nbsp;                |    |                atom   
 >zot\)\\z\<       |   9|            brnc   
&nbsp;                |  10|              piec   
&nbsp;                |    |                atom   
 >\\z\<           |  13|      piec   
&nbsp;                |    |        atom   
Required size 14 nodes
Starting second pass \(creation\)
 >\\A\(?&#8203;:\(?&#8203;:\)f\.\.\. |   1|  reg    
&nbsp;                |    |    brnc   
&nbsp;                |    |      piec   
&nbsp;                |    |        atom   
 >\(?&#8203;:\(?&#8203;:\)foo\.\.\. |   2|      piec   
&nbsp;                |    |        atom   
 >?&#8203;:\(?&#8203;:\)foo|\.\.\. |    |          reg    
 >\(?&#8203;:\)foo|ba\.\.\. |    |            brnc   
&nbsp;                |    |              piec   
&nbsp;                |    |                atom   
 >?&#8203;:\)foo|bar\.\.\. |    |                  reg    
 >\)foo|bar|z\.\.\. |    |                    brnc   
 >foo|bar|zo\.\.\. |   3|              piec   
&nbsp;                |    |                atom   
 >|bar|zot\)\\\.\.\. |   5|              tail~ NOTHING \(2\) \-> EXACT
&nbsp;                |    |            inst \- BRANCH
 >bar|zot\)\\z\<   |   6|            brnc   
&nbsp;                |   7|              piec   
&nbsp;                |    |                atom   
 >|zot\)\\z\<      |   9|            tail~ BRANCH \(2\) \-> BRANCH
 >zot\)\\z\<       |    |            brnc   
&nbsp;                |  10|              piec   
&nbsp;                |    |                atom   
 >\)\\z\<          |  12|            tail~ BRANCH \(6\) \-> BRANCH
&nbsp;                |  13|            tail~ BRANCH \(9\) \-> TAIL
&nbsp;                |    |            tsdy~ NOTHING \(3\) \-> PSEUDO
&nbsp;                |    |                ~ EXACT \<foo> \(4\) \-> EXACT
&nbsp;                |    |                ~ attach to TAIL \(12\) offset to 8
&nbsp;                |    |            tsdy~ EXACT \<bar> \(7\) \-> EXACT
&nbsp;                |    |                ~ attach to TAIL \(12\) offset to 5
&nbsp;                |    |            tsdy~ EXACT \<zot> \(10\) \-> EXACT
&nbsp;                |    |                ~ attach to TAIL \(12\) offset to 2
 >\\z\<           |    |      tail~ SBOL \(1\) \-> BRANCH
&nbsp;                |    |      piec   
&nbsp;                |    |        atom   
 >\<             |  14|      tail~ BRANCH \(2\)  
&nbsp;                |    |          ~ BRANCH \(6\)  
&nbsp;                |    |          ~ BRANCH \(9\)  
&nbsp;                |    |          ~ TAIL \(12\) \-> EOS
&nbsp;                |  15|    tail~ SBOL \(1\)  
&nbsp;                |    |        ~ BRANCH \(2\)  
&nbsp;                |    |        ~ BRANCH \(6\)  
&nbsp;                |    |        ~ BRANCH \(9\)  
&nbsp;                |    |        ~ TAIL \(12\)  
&nbsp;                |    |        ~ EOS \(13\) \-> END
first&#8203;:>  1&#8203;: SBOL \(2\)
first&#8203;:>  2&#8203;: BRANCH \(6\)
first at 2
Peep&#8203;:Pos&#8203;:0/0 Flags&#8203;: 0x0 Whilem\_c&#8203;: 0 Lcp&#8203;: 0 Last&#8203;:'' 0&#8203;:0/0 \*Fixed&#8203;:'' @&#8203; 0 Float&#8203;: '' @&#8203; 0/0
Peep>  1&#8203;: SBOL \(2\)
Peep&#8203;:Pos&#8203;:0/0 Flags&#8203;: 0x0 Whilem\_c&#8203;: 0 Lcp&#8203;: 0 Last&#8203;:'' 0&#8203;:0/0 \*Fixed&#8203;:'' @&#8203; 0 Float&#8203;: '' @&#8203; 0/0
Peep>  2&#8203;: BRANCH \(6\)
commit&#8203;: Pos&#8203;:0/0 Flags&#8203;: 0x0 Whilem\_c&#8203;: 0 Lcp&#8203;: 0 Last&#8203;:'' \-1&#8203;:0/0 \*Fixed&#8203;:'' @&#8203; 0 Float&#8203;: '' @&#8203; 0/0
&nbsp;  Peep&#8203;:Pos&#8203;:0/0 Flags&#8203;: 0x0 Whilem\_c&#8203;: 0 Lcp&#8203;: 0 
&nbsp;  Peep>  3&#8203;: NOTHING \(4\)
&nbsp;  Peep&#8203;:Pos&#8203;:0/0 Flags&#8203;: 0x0 Whilem\_c&#8203;: 0 Lcp&#8203;: 0 
&nbsp;  Peep>  4&#8203;: EXACT \<foo> \(12\)
&nbsp;    join>  4&#8203;: EXACT \<foo> \(12\)
&nbsp;    skip&#8203;:> 12&#8203;: TAIL \(13\)
&nbsp;  pre\-fin&#8203;:Pos&#8203;:0/0 Flags&#8203;: 0x0 Whilem\_c&#8203;: 0 Lcp&#8203;: 0 
&nbsp;  post\-fin&#8203;:Pos&#8203;:0/0 Flags&#8203;: 0x0 Whilem\_c&#8203;: 0 Lcp&#8203;: 0 
&nbsp;  Peep&#8203;:Pos&#8203;:0/0 Flags&#8203;: 0x0 Whilem\_c&#8203;: 0 Lcp&#8203;: 0 
&nbsp;  Peep>  7&#8203;: EXACT \<bar> \(12\)
&nbsp;    join>  7&#8203;: EXACT \<bar> \(12\)
&nbsp;    skip&#8203;:> 12&#8203;: TAIL \(13\)
&nbsp;  pre\-fin&#8203;:Pos&#8203;:0/0 Flags&#8203;: 0x0 Whilem\_c&#8203;: 0 Lcp&#8203;: 0 
&nbsp;  post\-fin&#8203;:Pos&#8203;:0/0 Flags&#8203;: 0x0 Whilem\_c&#8203;: 0 Lcp&#8203;: 0 
&nbsp;  Peep&#8203;:Pos&#8203;:0/0 Flags&#8203;: 0x0 Whilem\_c&#8203;: 0 Lcp&#8203;: 0 
&nbsp;  Peep> 10&#8203;: EXACT \<zot> \(12\)
&nbsp;    join> 10&#8203;: EXACT \<zot> \(12\)
&nbsp;    skip&#8203;:> 12&#8203;: TAIL \(13\)
&nbsp;  pre\-fin&#8203;:Pos&#8203;:0/0 Flags&#8203;: 0x0 Whilem\_c&#8203;: 0 Lcp&#8203;: 0 
&nbsp;  post\-fin&#8203;:Pos&#8203;:0/0 Flags&#8203;: 0x0 Whilem\_c&#8203;: 0 Lcp&#8203;: 0 
&nbsp;  Looking for TRIE'able sequences\. Tail node is&#8203;: EOS
&nbsp;  \- BRANCH \(2\) \-> NOTHING    => EXACT \<foo>    \(First==\-1\,Last==\-1\,Cur==2\)
&nbsp;  \- BRANCH \(6\) \-> EXACT \<bar>    => EOS    \(First==2\,Last==\-1\,Cur==6\)
&nbsp;  \- BRANCH \(9\) \-> EXACT \<zot>    => EOS    \(First==2\,Last==\-1\,Cur==9\)
&nbsp;  \- TAIL \(12\) \<SCAN FINISHED>
&nbsp;    make\_trie start==2\, first==2\, last==12\, tail==13 depth=1
&nbsp;    TRIE\(NATIVE\)&#8203;: W&#8203;:3 C&#8203;:6 Uq&#8203;:6 Min&#8203;:0 Max&#8203;:3
&nbsp;    Compiling trie using table compiler
&nbsp;      Char :    b   a   r   z   o   t
&nbsp;      State\+\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
&nbsp;         1 :    2   \.   \.   5   \.   \. \(   2\) W   1
&nbsp;         2 :    \.   3   \.   \.   \.   \. \(   1\)
&nbsp;         3 :    \.   \.   4   \.   \.   \. \(   1\)
&nbsp;         4 :    \.   \.   \.   \.   \.   \. \(   0\) W   2
&nbsp;         5 :    \.   \.   \.   \.   6   \. \(   1\)
&nbsp;         6 :    \.   \.   \.   \.   \.   7 \(   1\)
&nbsp;         7 :    \.   \.   \.   \.   \.   \. \(   0\) W   3
&nbsp;    Alloc&#8203;: 43 Orig&#8203;: 43 elements\, Final&#8203;:6\. Savings of %86\.05
&nbsp;    Statecount&#8203;:8 Lasttrans&#8203;:7
&nbsp;      Char : Match Base  Ofs     b   a   r   z   o   t
&nbsp;      State|\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
&nbsp;      \#   1| W   1 @&#8203;   6 \+ 0\[    2   \.   \.   5   \.   \.\]
&nbsp;      \#   2|       @&#8203;   6 \+ 1\[    \.   3   \.   \.   \.   \.\]
&nbsp;      \#   3|       @&#8203;   6 \+ 2\[    \.   \.   4   \.   \.   \.\]
&nbsp;      \#   4| W   2 @&#8203;   0 
&nbsp;      \#   5|       @&#8203;   6 \+ 4\[    \.   \.   \.   \.   6   \.\]
&nbsp;      \#   6|       @&#8203;   6 \+ 5\[    \.   \.   \.   \.   \.   7\]
&nbsp;      \#   7| W   3 @&#8203;   0 
&nbsp;    MJD offset&#8203;:8 MJD length&#8203;:0
Peep&#8203;:Pos&#8203;:3/0 Flags&#8203;: 0x0 Whilem\_c&#8203;: 0 Lcp&#8203;: 0 Last&#8203;:'' \-1&#8203;:0/0 \*Fixed&#8203;:'' @&#8203; 0 Float&#8203;: '' @&#8203; 0/0
Peep> 12&#8203;: TAIL \(13\)
Peep&#8203;:Pos&#8203;:3/0 Flags&#8203;: 0x0 Whilem\_c&#8203;: 0 Lcp&#8203;: 0 Last&#8203;:'' \-1&#8203;:0/0 \*Fixed&#8203;:'' @&#8203; 0 Float&#8203;: '' @&#8203; 0/0
Peep> 13&#8203;: EOS \(14\)
pre\-fin&#8203;:Pos&#8203;:3/0 Flags&#8203;: 0x1 Whilem\_c&#8203;: 0 Lcp&#8203;: 0 Last&#8203;:'' \-1&#8203;:0/0 \*Fixed&#8203;:'' @&#8203; 0 Float&#8203;: '' @&#8203; 0/0
post\-fin&#8203;:Pos&#8203;:3/0 Flags&#8203;: 0x1 Whilem\_c&#8203;: 0 Lcp&#8203;: 0 Last&#8203;:'' \-1&#8203;:0/0 \*Fixed&#8203;:'' @&#8203; 0 Float&#8203;: '' @&#8203; 0/0
commit&#8203;: Pos&#8203;:3/0 Flags&#8203;: 0x4 Whilem\_c&#8203;: 0 Lcp&#8203;: 0 Last&#8203;:'' \-1&#8203;:0/0 \*Fixed&#8203;:'' @&#8203; 3 Float&#8203;: '' @&#8203; 0/0
minlen&#8203;: 3 r\->minlen&#8203;:0
Final program&#8203;:
&nbsp;   1&#8203;: SBOL \(2\)
&nbsp;   2&#8203;: TRIE\-EXACT\<S&#8203;:1/7 W&#8203;:3 L&#8203;:0/3 C&#8203;:6/6>\[bz\] \(13\)
&nbsp;      \<> \(4\)
&nbsp;   4&#8203;:   EXACT \<foo> \(13\)
&nbsp;      \<bar> \(13\)
&nbsp;      \<zot> \(13\)
&nbsp;  13&#8203;: EOS \(14\)
&nbsp;  14&#8203;: END \(0\)
anchored ""$ at 3 anchored\(SBOL\) minlen 3 
r\->extflags&#8203;: ANCH\_SBOL 
Matching REx "\\A\(?&#8203;:\(?&#8203;:\)foo|bar|zot\)\\z" against "foo"
&nbsp;  Setting an EVAL scope\, savestack=3
regmatch start
&nbsp;   0 \<> \<foo>                |  1&#8203;:SBOL\(2\)
&nbsp;   0 \<> \<foo>                |  2&#8203;:TRIE\-EXACT\<S&#8203;:1/7 W&#8203;:3 L&#8203;:0/3 C&#8203;:6/6>\[bz\]\(13\)
&nbsp;                                  matched empty string\.\.\.
&nbsp;   0 \<> \<foo>                | 13&#8203;:EOS\(14\)
&nbsp;                                  failed\.\.\.
Match failed
FAIL
Freeing REx&#8203;: "\\A\(?&#8203;:\(?&#8203;:\)foo|bar|zot\)\\z"

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=medium

Site configuration information for perl 5.12.2:

Configured by ilmari at Thu Sep  9 18:09:40 BST 2010.

Summary of my perl5 (revision 5 version 12 subversion 2) configuration:
  Commit id: 7a3b65c9d99f69553fffe01f73d49fe9abf95120
  Platform:
    osname=linux, osvers=2.6.32-24-generic, archname=x86_64-linux
    uname='linux fenchurch 2.6.32-24-generic #42-ubuntu smp fri aug 20 14:21:58 utc 2010 x86_64 gnulinux '
    config_args='-Dprefix=/home/ilmari/perls/v5.12.2 -Duse64bitint -Uuselongdouble -DDEBUGGING -Doptimize=-g3 -Dusedevel -Uusethreads -Uversiononly -Dcccdlflags=-fPIC -Duseshrplib -des'
    hint=recommended, useposix=true, d_sigaction=define
    useithreads=undef, usemultiplicity=undef
    useperlio=define, d_sfio=undef, uselargefiles=define, usesocks=undef
    use64bitint=define, use64bitall=define, uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='cc', ccflags ='-DDEBUGGING -fno-strict-aliasing -pipe -fstack-protector -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64',
    optimize='-g3',
    cppflags='-DDEBUGGING -fno-strict-aliasing -pipe -fstack-protector -I/usr/local/include'
    ccversion='', gccversion='4.4.3', gccosandvers=''
    intsize=4, longsize=8, ptrsize=8, doublesize=8, byteorder=12345678
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=16
    ivtype='long', ivsize=8, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=8
    alignbytes=8, prototype=define
  Linker and Libraries:
    ld='cc', ldflags =' -fstack-protector -L/usr/local/lib'
    libpth=/usr/local/lib /lib /usr/lib /lib64 /usr/lib64
    libs=-lnsl -lgdbm -ldb -ldl -lm -lcrypt -lutil -lc -lgdbm_compat
    perllibs=-lnsl -ldl -lm -lcrypt -lutil -lc
    libc=/lib/libc-2.11.1.so, so=so, useshrplib=true, libperl=libperl.so
    gnulibc_version='2.11.1'
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-Wl,-E -Wl,-rpath,/home/ilmari/perls/v5.12.2/lib/5.12.2/x86_64-linux/CORE'
    cccdlflags='-fPIC', lddlflags='-shared -g3 -L/usr/local/lib -fstack-protector'

Locally applied patches:
    


@INC for perl 5.12.2:
    /home/ilmari/perls/v5.12.2/lib/site_perl/5.12.2/x86_64-linux
    /home/ilmari/perls/v5.12.2/lib/site_perl/5.12.2
    /home/ilmari/perls/v5.12.2/lib/5.12.2/x86_64-linux
    /home/ilmari/perls/v5.12.2/lib/5.12.2
    .


Environment for perl 5.12.2:
    HOME=/home/ilmari
    LANG=en_GB.utf8
    LANGUAGE (unset)
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=/home/ilmari/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/Babel/scripts:/home/ilmari/Babel/hk/script:/home/ilmari/Babel/hk/logfilters:/home/ilmari/Babel/hk/markf-svn
    PERL_AUTOINSTALL_PREFER_CPAN=1
    PERL_BADLANG (unset)
    PERL_CPANM_DEV=1
    PERL_MB_OPT=--installdirs site
    PERL_MM_OPT=INSTALLDIRS=site
    PERL_MM_USE_DEFAULT=1
    SHELL=/bin/bash

```
</details>

**Comment by p5pRT** [comment]
### From @cpansprout

On Wed Oct 13 02&#8203;:53&#8203;:36 2010\, ilmari wrote&#8203;:
> $ perl5\.8\.8 \-le 'print "foo" =~ /\\A\(?&#8203;:\(?&#8203;:\)foo|bar|zot\)\\z/ ? "match" :
> "FAIL"'
> match
> 
> $ perl5\.12\.2 \-le 'print "foo" =~ /\\A\(?&#8203;:\(?&#8203;:\)foo|bar|zot\)\\z/ ? "match" :
> "FAIL"'
> FAIL
> 
> \.\.\.
> 
> This regression was introduced in perl 5\.9\.5\.

by&#8203;:

From&#8203;: Yves Orton \<demerphq gmail\.com>
Date&#8203;: Sat\, 2 Sep 2006 16&#8203;:40&#8203;:12 \+0000 \(\+0200\)
Subject&#8203;: Re&#8203;: \[PATCH\] Trie jumping
X\-Git\-Tag&#8203;: perl\-5\.9\.5~2170
X\-Git\-Url&#8203;:
http&#8203;://perl5\.git\.perl\.org/perl\.git/commitdiff\_plain/786e8c118e1218e4c348fecf469934e080881633

Re&#8203;: \[PATCH\] Trie jumping
Message\-ID&#8203;: \<9b18b3110609020740y2eb9004cpab313c3353a437ca@&#8203;mail\.gmail\.com>

p4raw\-id&#8203;: //depot/perl@&#8203;28785


See also
\<http&#8203;://www\.nntp\.perl\.org/group/perl\.perl5\.porters/2006/09/msg116095\.html>\.

**Comment by p5pRT** [comment]
### From [Unknown Contact. See original ticket]

On Wed Oct 13 02&#8203;:53&#8203;:36 2010\, ilmari wrote&#8203;:
> $ perl5\.8\.8 \-le 'print "foo" =~ /\\A\(?&#8203;:\(?&#8203;:\)foo|bar|zot\)\\z/ ? "match" :
> "FAIL"'
> match
> 
> $ perl5\.12\.2 \-le 'print "foo" =~ /\\A\(?&#8203;:\(?&#8203;:\)foo|bar|zot\)\\z/ ? "match" :
> "FAIL"'
> FAIL
> 
> \.\.\.
> 
> This regression was introduced in perl 5\.9\.5\.

by&#8203;:

From&#8203;: Yves Orton \<demerphq gmail\.com>
Date&#8203;: Sat\, 2 Sep 2006 16&#8203;:40&#8203;:12 \+0000 \(\+0200\)
Subject&#8203;: Re&#8203;: \[PATCH\] Trie jumping
X\-Git\-Tag&#8203;: perl\-5\.9\.5~2170
X\-Git\-Url&#8203;:
http&#8203;://perl5\.git\.perl\.org/perl\.git/commitdiff\_plain/786e8c118e1218e4c348fecf469934e080881633

Re&#8203;: \[PATCH\] Trie jumping
Message\-ID&#8203;: \<9b18b3110609020740y2eb9004cpab313c3353a437ca@&#8203;mail\.gmail\.com>

p4raw\-id&#8203;: //depot/perl@&#8203;28785


See also
\<http&#8203;://www\.nntp\.perl\.org/group/perl\.perl5\.porters/2006/09/msg116095\.html>\.

**Comment by p5pRT** [comment]
@cpansprout - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @cpansprout

On Sun Oct 17 17&#8203;:42&#8203;:17 2010\, sprout wrote&#8203;:
> On Wed Oct 13 02&#8203;:53&#8203;:36 2010\, ilmari wrote&#8203;:
> > This regression was introduced in perl 5\.9\.5\.
> 
> by&#8203;:
> 
> From&#8203;: Yves Orton \<demerphq gmail\.com>
> Date&#8203;: Sat\, 2 Sep 2006 16&#8203;:40&#8203;:12 \+0000 \(\+0200\)
> Subject&#8203;: Re&#8203;: \[PATCH\] Trie jumping
> X\-Git\-Tag&#8203;: perl\-5\.9\.5~2170
> X\-Git\-Url&#8203;:
>
http&#8203;://perl5\.git\.perl\.org/perl\.git/commitdiff\_plain/786e8c118e1218e4c348fecf469934e080881633
> 
> Re&#8203;: \[PATCH\] Trie jumping
> Message\-ID&#8203;:
> \<9b18b3110609020740y2eb9004cpab313c3353a437ca@&#8203;mail\.gmail\.com>
> 
> p4raw\-id&#8203;: //depot/perl@&#8203;28785

and fixed by 20dbff7c22\.

**Comment by p5pRT** [comment]
@cpansprout - Status changed from 'open' to 'resolved'
